File: sdk/go/stepauth/envelope_test.go
package stepauth

import (
	"bytes"
	"crypto/ed25519"
	"crypto/rand"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"testing"

	"github.com/cloudflare/circl/sign/mldsa/mldsa65"
)

func testKeyset(t *testing.T) *SigningKeyset {
	t.Helper()
	ks, err := GenerateKeyset("ks_test")
	if err != nil {
		t.Fatalf("generating keyset: %v", err)
	}
	return ks
}

func reEncode(t *testing.T, envelopeJSON []byte, fn func(*Envelope)) []byte {
	t.Helper()
	var env Envelope
	if err := json.Unmarshal(envelopeJSON, &env); err != nil {
		t.Fatalf("unmarshaling envelope: %v", err)
	}
	fn(&env)
	out, err := json.Marshal(env)
	if err != nil {
		t.Fatalf("marshaling envelope: %v", err)
	}
	return out
}

func TestSigningKeysetStringRedactsPrivateKeys(t *testing.T) {
	ks := testKeyset(t)

	out := fmt.Sprintf("%+v keys:%+v", ks, ks.Keys)

	for _, k := range ks.Keys {
		if strings.Contains(out, fmt.Sprintf("%v", k.Priv)) {
			t.Errorf("formatted output contains raw private key bytes: %s", out)
		}
	}
	if !strings.Contains(out, "REDACTED") {
		t.Errorf("formatted output does not redact private key material: %s", out)
	}
}

func TestEnvelopeRoundTrip(t *testing.T) {
	ks := testKeyset(t)
	payload := []byte(`{"requestId":"req_1","senderId":"sp.example.com"}`)

	envelopeJSON, err := SignEnvelope(ks, payload)
	if err != nil {
		t.Fatalf("signing: %v", err)
	}

	got, err := VerifyEnvelope(envelopeJSON, []Keyset{ks.PublicKeyset()})
	if err != nil {
		t.Fatalf("verifying: %v", err)
	}
	if !bytes.Equal(got, payload) {
		t.Errorf("payload round-tripped as %q, want %q", got, payload)
	}
}

// TestKeysetCarriesBothAlgorithms guards the hybrid pair itself: dropping a
// component would leave signatures verifying against one primitive alone.
func TestKeysetCarriesBothAlgorithms(t *testing.T) {
	ks := testKeyset(t)
	wantAlgs := []string{AlgEd25519, AlgMLDSA65}

	if len(ks.Keys) != len(wantAlgs) {
		t.Fatalf("keyset has %d components, want %d", len(ks.Keys), len(wantAlgs))
	}
	for i, alg := range wantAlgs {
		if ks.Keys[i].Alg != alg {
			t.Errorf("component %d is %q, want %q", i, ks.Keys[i].Alg, alg)
		}
		if ks.Keys[i].Idx != i {
			t.Errorf("component %d carries idx %d", i, ks.Keys[i].Idx)
		}
	}
}

// TestSeedDerivationIsDeterministic checks that one seed always expands to
// the same public key.
func TestSeedDerivationIsDeterministic(t *testing.T) {
	seeds := map[string][]byte{
		AlgEd25519: bytes.Repeat([]byte{0x01}, ed25519.SeedSize),
		AlgMLDSA65: bytes.Repeat([]byte{0x02}, mldsa65.SeedSize),
	}

	for alg, seed := range seeds {
		t.Run(alg, func(t *testing.T) {
			first, err := KeyFromSeed(alg, seed)
			if err != nil {
				t.Fatalf("expanding seed: %v", err)
			}
			second, err := KeyFromSeed(alg, seed)
			if err != nil {
				t.Fatalf("expanding seed again: %v", err)
			}
			if !bytes.Equal(first, second) {
				t.Error("one seed expanded to two different public keys")
			}
			if len(first) == 0 {
				t.Error("seed expanded to an empty public key")
			}
		})
	}
}

// TestKeysetJSONIsBackwardCompatible pins the persisted shape. ML-DSA private
// keys cannot be reduced back to a seed, so a field rename here would strand
// every existing keyset with no migration path short of re-registering with
// the hub.
func TestKeysetJSONIsBackwardCompatible(t *testing.T) {
	raw, err := json.Marshal(testKeyset(t))
	if err != nil {
		t.Fatalf("marshaling keyset: %v", err)
	}

	var shape struct {
		KeysetID string `json:"keysetId"`
		Keys     []struct {
			Idx  int    `json:"idx"`
			Alg  string `json:"alg"`
			Pub  string `json:"pub"`
			Priv string `json:"priv"`
		} `json:"keys"`
	}
	if err := json.Unmarshal(raw, &shape); err != nil {
		t.Fatalf("keyset JSON does not match the persisted shape: %v", err)
	}

	if shape.KeysetID == "" || len(shape.Keys) != len(KeysetAlgs()) {
		t.Fatalf("unexpected keyset shape: %s", raw)
	}
	for i, k := range shape.Keys {
		if k.Alg != KeysetAlgs()[i] || k.Idx != i || k.Pub == "" || k.Priv == "" {
			t.Errorf("component %d did not round-trip: %+v", i, k)
		}
	}

	// A stored keyset must come back able to sign.
	var restored SigningKeyset
	if err := json.Unmarshal(raw, &restored); err != nil {
		t.Fatalf("restoring keyset: %v", err)
	}
	envelopeJSON, err := SignEnvelope(&restored, []byte(`{"requestId":"req_1"}`))
	if err != nil {
		t.Fatalf("signing with a restored keyset: %v", err)
	}
	if _, err := VerifyEnvelope(envelopeJSON, []Keyset{restored.PublicKeyset()}); err != nil {
		t.Errorf("envelope from a restored keyset did not verify: %v", err)
	}
}

func TestVerifyRejectsTamperedPayload(t *testing.T) {
	ks := testKeyset(t)
	envelopeJSON, err := SignEnvelope(ks, []byte(`{"requestId":"req_1"}`))
	if err != nil {
		t.Fatalf("signing: %v", err)
	}

	tampered := reEncode(t, envelopeJSON, func(env *Envelope) {
		env.Payload = base64.StdEncoding.EncodeToString([]byte(`{"requestId":"req_2"}`))
	})

	if _, err := VerifyEnvelope(tampered, []Keyset{ks.PublicKeyset()}); !errors.Is(err, ErrInvalidSignature) {
		t.Errorf("tampered payload verified, err = %v", err)
	}
}

func TestVerifyRejectsUnknownKeyset(t *testing.T) {
	ks := testKeyset(t)
	envelopeJSON, err := SignEnvelope(ks, []byte(`{}`))
	if err != nil {
		t.Fatalf("signing: %v", err)
	}

	other := ks.PublicKeyset()
	other.KeysetID = "ks_someone_else"

	if _, err := VerifyEnvelope(envelopeJSON, []Keyset{other}); !errors.Is(err, ErrInvalidSignature) {
		t.Errorf("envelope verified against an unresolvable keysetId, err = %v", err)
	}
}

// TestVerifyRejectsZeroKeyKeyset covers the vacuous all-of: with no components
// every check passes trivially and any payload verifies.
func TestVerifyRejectsZeroKeyKeyset(t *testing.T) {
	// Built directly, not via SignEnvelope: a zero-signature envelope keeps
	// the signature-count check (0 == 0) from rejecting first, so only the
	// zero-key guard can catch it.
	envelopeJSON, err := json.Marshal(Envelope{
		Payload:   base64.StdEncoding.EncodeToString([]byte(`{}`)),
		Signature: Signature{KeysetID: "ks_test", Signatures: []string{}},
	})
	if err != nil {
		t.Fatalf("marshaling envelope: %v", err)
	}

	empty := Keyset{KeysetID: "ks_test"}
	if _, err := VerifyEnvelope(envelopeJSON, []Keyset{empty}); !errors.Is(err, ErrInvalidSignature) {
		t.Errorf("zero-key keyset verified an envelope, err = %v", err)
	}
}

func TestVerifyRejectsSignatureCountMismatch(t *testing.T) {
	ks := testKeyset(t)
	envelopeJSON, err := SignEnvelope(ks, []byte(`{}`))
	if err != nil {
		t.Fatalf("signing: %v", err)
	}

	for name, mutate := range map[string]func(*Envelope){
		"too few":  func(env *Envelope) { env.Signature.Signatures = env.Signature.Signatures[:1] },
		"too many": func(env *Envelope) { env.Signature.Signatures = append(env.Signature.Signatures, "AA==") },
	} {
		t.Run(name, func(t *testing.T) {
			mutated := reEncode(t, envelopeJSON, mutate)
			if _, err := VerifyEnvelope(mutated, []Keyset{ks.PublicKeyset()}); !errors.Is(err, ErrInvalidSignature) {
				t.Errorf("%s signatures verified, err = %v", name, err)
			}
		})
	}
}

// TestVerifyRejectsSingleBadComponent is the hybrid downgrade check. An
// implementation that stops at the first valid signature, or only checks the
// classical half, passes every other test in this file and fails this one.
func TestVerifyRejectsSingleBadComponent(t *testing.T) {
	ks := testKeyset(t)
	payload := []byte(`{"requestId":"req_1"}`)

	envelopeJSON, err := SignEnvelope(ks, payload)
	if err != nil {
		t.Fatalf("signing: %v", err)
	}

	// A signature valid for other bytes: structurally sound, wrong message.
	decoy, err := SignEnvelope(ks, []byte(`{"requestId":"req_other"}`))
	if err != nil {
		t.Fatalf("signing decoy: %v", err)
	}
	var decoyEnv Envelope
	if err := json.Unmarshal(decoy, &decoyEnv); err != nil {
		t.Fatalf("unmarshaling decoy: %v", err)
	}

	for idx, alg := range KeysetAlgs() {
		t.Run(alg, func(t *testing.T) {
			swapped := reEncode(t, envelopeJSON, func(env *Envelope) {
				env.Signature.Signatures[idx] = decoyEnv.Signature.Signatures[idx]
			})
			if _, err := VerifyEnvelope(swapped, []Keyset{ks.PublicKeyset()}); !errors.Is(err, ErrInvalidSignature) {
				t.Errorf("envelope verified with only the %s component wrong, err = %v", alg, err)
			}
		})
	}
}

func TestSignRejectsDuplicateAlgKeyset(t *testing.T) {
	pub1, priv1, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		t.Fatalf("generating key: %v", err)
	}
	pub2, priv2, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		t.Fatalf("generating key: %v", err)
	}
	ks := &SigningKeyset{
		KeysetID: "ks_test",
		Keys: []SigningKey{
			{Idx: 0, Alg: AlgEd25519, Pub: pub1, Priv: priv1},
			{Idx: 1, Alg: AlgEd25519, Pub: pub2, Priv: priv2},
		},
	}

	if _, err := SignEnvelope(ks, []byte(`{}`)); err == nil {
		t.Error("signing with a duplicate-algorithm keyset succeeded")
	}
}

func TestVerifyRejectsDuplicateAlgKeyset(t *testing.T) {
	pub1, _, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		t.Fatalf("generating key: %v", err)
	}
	pub2, _, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		t.Fatalf("generating key: %v", err)
	}
	keyset := Keyset{
		KeysetID: "ks_test",
		Keys: []KeysetKey{
			{Idx: 0, Alg: AlgEd25519, Pub: base64.StdEncoding.EncodeToString(pub1)},
			{Idx: 1, Alg: AlgEd25519, Pub: base64.StdEncoding.EncodeToString(pub2)},
		},
	}
	envelopeJSON, err := json.Marshal(Envelope{
		Payload:   base64.StdEncoding.EncodeToString([]byte(`{}`)),
		Signature: Signature{KeysetID: "ks_test", Signatures: []string{"AA==", "AA=="}},
	})
	if err != nil {
		t.Fatalf("marshaling envelope: %v", err)
	}

	if _, err := VerifyEnvelope(envelopeJSON, []Keyset{keyset}); !errors.Is(err, ErrInvalidSignature) {
		t.Errorf("duplicate-algorithm keyset verified an envelope, err = %v", err)
	}
}

// TestEnvelopeWireEncoding pins the two things a language port can silently
// get wrong: the base64 alphabet and the JSON key names. Both are invisible
// to a test that signs and verifies with the same code.
func TestEnvelopeWireEncoding(t *testing.T) {
	ks, err := GenerateKeyset("ks_wire")
	if err != nil {
		t.Fatalf("generating keyset: %v", err)
	}

	// Encodes to "/++++Q==" in the standard alphabet and "_----Q==" in
	// base64url, so the two are distinguishable here where short ASCII
	// payloads make them identical.
	payload := []byte{0xff, 0xef, 0xbe, 0xfb, 0xff, 0x41}
	wantPayload := base64.StdEncoding.EncodeToString(payload)

	envelopeBytes, err := SignEnvelope(ks, payload)
	if err != nil {
		t.Fatalf("signing: %v", err)
	}

	var raw map[string]json.RawMessage
	if err := json.Unmarshal(envelopeBytes, &raw); err != nil {
		t.Fatalf("unmarshaling envelope: %v", err)
	}
	if _, ok := raw["payload"]; !ok {
		t.Error(`envelope has no "payload" key`)
	}
	if _, ok := raw["signature"]; !ok {
		t.Error(`envelope has no "signature" key`)
	}

	var got struct {
		Payload   string `json:"payload"`
		Signature struct {
			KeysetID   string   `json:"keysetId"`
			Signatures []string `json:"signatures"`
		} `json:"signature"`
	}
	if err := json.Unmarshal(envelopeBytes, &got); err != nil {
		t.Fatalf("unmarshaling envelope: %v", err)
	}
	if got.Payload != wantPayload {
		t.Errorf("payload encoding = %q, want %q (standard alphabet)", got.Payload, wantPayload)
	}
	if got.Signature.KeysetID != "ks_wire" {
		t.Errorf(`signature.keysetId = %q, want "ks_wire"`, got.Signature.KeysetID)
	}
	if len(got.Signature.Signatures) != len(KeysetAlgs()) {
		t.Errorf("signature.signatures = %d entries, want %d", len(got.Signature.Signatures), len(KeysetAlgs()))
	}
}

func TestDecodePayloadDoesNotVerify(t *testing.T) {
	ks := testKeyset(t)
	envelopeJSON, err := SignEnvelope(ks, []byte(`{"requestId":"req_1"}`))
	if err != nil {
		t.Fatalf("signing: %v", err)
	}

	tampered := reEncode(t, envelopeJSON, func(env *Envelope) {
		env.Payload = base64.StdEncoding.EncodeToString([]byte(`{"requestId":"forged"}`))
	})

	// Routing needs the payload before the keyset is known, so this must decode
	// what VerifyEnvelope would reject.
	got, err := DecodeEnvelopePayload(tampered)
	if err != nil {
		t.Fatalf("decoding: %v", err)
	}
	if string(got) != `{"requestId":"forged"}` {
		t.Errorf("decoded %q", got)
	}
}

File: sdk/es/test/envelope.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";

import {
  ALG_ED25519,
  ALG_ML_DSA_65,
  KEYSET_ALGS,
  InvalidSignatureError,
  EnvelopeFormatError,
  generateKeyset,
  keyFromSeed,
  publicKeyset,
  signingKeysetToJSON,
  signingKeysetFromJSON,
  signEnvelope,
  verifyEnvelope,
  decodeEnvelopePayload,
  type Envelope,
  type SigningKeyset,
} from "../src/stepauth/index.js";
import { signComponent, verifyComponent } from "../src/stepauth/crypto.js";

const b64 = (s: string) => Buffer.from(s).toString("base64");

function testKeyset(): SigningKeyset {
  return generateKeyset("ks_test");
}

function reEncode(envelopeJSON: string, fn: (env: Envelope) => void): string {
  const env = JSON.parse(envelopeJSON) as Envelope;
  fn(env);
  return JSON.stringify(env);
}

test("envelope round trip", () => {
  const ks = testKeyset();
  const payload = new TextEncoder().encode(`{"requestId":"req_1"}`);

  const envelopeJSON = signEnvelope(ks, payload);
  const got = verifyEnvelope(envelopeJSON, [publicKeyset(ks)]);

  assert.deepEqual(Buffer.from(got), Buffer.from(payload));
});

test("keyset carries both algorithms", () => {
  const ks = testKeyset();

  assert.equal(ks.keys.length, KEYSET_ALGS.length);
  KEYSET_ALGS.forEach((alg, i) => {
    assert.equal(ks.keys[i].alg, alg);
    assert.equal(ks.keys[i].idx, i);
  });
});

// What the shared keygen vectors rely on: one seed expands to one public key,
// in every implementation of every language.
test("seed derivation is deterministic", () => {
  for (const [alg, len] of [
    [ALG_ED25519, 32],
    [ALG_ML_DSA_65, 32],
  ] as const) {
    const seed = new Uint8Array(len).fill(7);
    const first = keyFromSeed(alg, seed);
    const second = keyFromSeed(alg, seed);
    assert.deepEqual(Buffer.from(first), Buffer.from(second), alg);
    assert.ok(first.length > 0, alg);
  }
});

// A keyset object handed to console.log or an uncareful JSON.stringify call
// must not spill its private key material.
test("generateKeyset redacts private key material from JSON.stringify", () => {
  const ks = testKeyset();
  assert.ok(!JSON.stringify(ks).includes('"priv"'));
});

// Pins the persisted shape. SPs already hold serialized keysets in this form,
// and ML-DSA private keys cannot be reduced back to a seed, so a field rename
// here strands every existing keyset.
test("keyset JSON is backward compatible", () => {
  const ks = testKeyset();
  const stored = signingKeysetToJSON(ks) as {
    keysetId: string;
    keys: { idx: number; alg: string; pub: string; priv: string }[];
  };

  assert.equal(stored.keys.length, KEYSET_ALGS.length);
  stored.keys.forEach((k, i) => {
    assert.equal(k.alg, KEYSET_ALGS[i]);
    assert.equal(k.idx, i);
    assert.ok(k.pub.length > 0 && k.priv.length > 0);
  });

  // A stored keyset must come back able to sign.
  const restored = signingKeysetFromJSON(stored);
  const envelopeJSON = signEnvelope(restored, new TextEncoder().encode(`{"a":1}`));
  assert.ok(verifyEnvelope(envelopeJSON, [publicKeyset(restored)]));
});

test("keyset round-trips through JSON.stringify/parse and still signs", () => {
  const ks = testKeyset();
  const json = JSON.stringify(signingKeysetToJSON(ks));
  const restored = signingKeysetFromJSON(JSON.parse(json));

  const envelopeJSON = signEnvelope(restored, new TextEncoder().encode(`{"a":1}`));
  assert.ok(verifyEnvelope(envelopeJSON, [publicKeyset(restored)]));
});

test("signingKeysetFromJSON rejects a malformed object", () => {
  for (const bad of [null, {}, { keysetId: "ks" }, { keysetId: "ks", keys: [{}] }]) {
    assert.throws(() => signingKeysetFromJSON(bad), /stepauth:/);
  }
});

test("verify rejects a tampered payload", () => {
  const ks = testKeyset();
  const envelopeJSON = signEnvelope(ks, new TextEncoder().encode(`{"requestId":"req_1"}`));

  const tampered = reEncode(envelopeJSON, (env) => {
    env.payload = b64(`{"requestId":"req_2"}`);
  });

  assert.throws(
    () => verifyEnvelope(tampered, [publicKeyset(ks)]),
    InvalidSignatureError,
  );
});

test("verify rejects an unresolvable keysetId", () => {
  const ks = testKeyset();
  const envelopeJSON = signEnvelope(ks, new TextEncoder().encode("{}"));

  const other = publicKeyset(ks);
  other.keysetId = "ks_someone_else";

  assert.throws(() => verifyEnvelope(envelopeJSON, [other]), InvalidSignatureError);
});

// The vacuous all-of: with no components every check passes trivially and any
// payload verifies. signatures is also empty so the count check (0 === 0)
// clears without exercising the guard under test.
test("verify rejects a zero-key keyset", () => {
  const envelopeJSON = JSON.stringify({
    payload: b64("{}"),
    signature: { keysetId: "ks_test", signatures: [] },
  });

  assert.throws(
    () => verifyEnvelope(envelopeJSON, [{ keysetId: "ks_test", keys: [] }]),
    InvalidSignatureError,
  );
});

test("verify rejects a signature count mismatch", () => {
  const ks = testKeyset();
  const envelopeJSON = signEnvelope(ks, new TextEncoder().encode("{}"));

  const tooFew = reEncode(envelopeJSON, (env) => {
    env.signature.signatures = env.signature.signatures.slice(0, 1);
  });
  const tooMany = reEncode(envelopeJSON, (env) => {
    env.signature.signatures = [...env.signature.signatures, "AA=="];
  });

  for (const [name, mutated] of [
    ["too few", tooFew],
    ["too many", tooMany],
  ] as const) {
    assert.throws(
      () => verifyEnvelope(mutated, [publicKeyset(ks)]),
      InvalidSignatureError,
      name,
    );
  }
});

// The hybrid downgrade check. An implementation that stops at the first valid
// signature, or only checks the classical half, passes every other test in this
// file and fails this one.
test("verify rejects a single bad component", () => {
  const ks = testKeyset();
  const payload = new TextEncoder().encode(`{"requestId":"req_1"}`);
  const envelopeJSON = signEnvelope(ks, payload);

  // A signature valid for other bytes: structurally sound, wrong message.
  const decoy = JSON.parse(
    signEnvelope(ks, new TextEncoder().encode(`{"requestId":"req_other"}`)),
  ) as Envelope;

  KEYSET_ALGS.forEach((alg, idx) => {
    const swapped = reEncode(envelopeJSON, (env) => {
      env.signature.signatures[idx] = decoy.signature.signatures[idx]!;
    });
    assert.throws(
      () => verifyEnvelope(swapped, [publicKeyset(ks)]),
      InvalidSignatureError,
      `only the ${alg} component was wrong`,
    );
  });
});

// The hybrid pair is mandatory in index order: a keyset missing a component,
// or repeating one, must never sign or verify.
test("signEnvelope and verifyEnvelope reject a keyset that isn't the exact v1 pair in order", () => {
  const ks = testKeyset();
  const classicalOnly = { keysetId: ks.keysetId, keys: [ks.keys[0]] };
  const duplicated = { keysetId: ks.keysetId, keys: [ks.keys[0], ks.keys[0]] };

  for (const bad of [classicalOnly, duplicated]) {
    assert.throws(() => signEnvelope(bad, new TextEncoder().encode("{}")));
  }

  const validEnvelope = signEnvelope(ks, new TextEncoder().encode("{}"));
  const badKeysets = [publicKeyset(classicalOnly), publicKeyset(duplicated)];
  for (const keyset of badKeysets) {
    assert.throws(() => verifyEnvelope(validEnvelope, [keyset]), InvalidSignatureError);
  }
});

test("decoding a payload does not verify it", () => {
  const ks = testKeyset();
  const envelopeJSON = signEnvelope(ks, new TextEncoder().encode(`{"requestId":"req_1"}`));

  const tampered = reEncode(envelopeJSON, (env) => {
    env.payload = b64(`{"requestId":"forged"}`);
  });

  // Routing needs the payload before the keyset is known, so this must decode
  // what verifyEnvelope would reject.
  const got = decodeEnvelopePayload(tampered);
  assert.equal(new TextDecoder().decode(got), `{"requestId":"forged"}`);
});

test("verify rejects lenient base64 the hub would reject", () => {
  const ks = testKeyset();
  const envelopeJSON = signEnvelope(ks, new TextEncoder().encode("{}"));

  for (const mutate of [
    (env: Envelope) => {
      env.payload = env.payload.replace(/=+$/, ""); // padding stripped
    },
    (env: Envelope) => {
      env.payload = env.payload + "@@@";
    },
    (env: Envelope) => {
      env.payload = env.payload.slice(0, 2) + " " + env.payload.slice(2);
    },
    (env: Envelope) => {
      env.payload = "q83v_-7dqw"; // base64url alphabet
    },
  ]) {
    const mutated = reEncode(envelopeJSON, mutate);
    assert.throws(() => verifyEnvelope(mutated, [publicKeyset(ks)]), InvalidSignatureError);
  }
});

test("verify accepts base64 forms Go's StdEncoding accepts", () => {
  const ks = testKeyset();

  // Line-wrapped: decoding ignores embedded CR/LF, so the signed bytes are
  // unchanged and the signature still verifies.
  const wrapped = signEnvelope(ks, new TextEncoder().encode(`{"requestId":"req_1"}`));
  const lineWrapped = reEncode(wrapped, (env) => {
    env.payload = env.payload.slice(0, 4) + "\n" + env.payload.slice(4);
  });
  assert.doesNotThrow(() => verifyEnvelope(lineWrapped, [publicKeyset(ks)]));

  // "AB==" has non-canonical unused trailing bits but decodes to the same
  // single zero byte as the canonical "AA==" this payload was signed as.
  const zeroByte = signEnvelope(ks, new Uint8Array([0]));
  const noncanonical = reEncode(zeroByte, (env) => {
    env.payload = "AB==";
  });
  assert.doesNotThrow(() => verifyEnvelope(noncanonical, [publicKeyset(ks)]));
});

test("decodeEnvelopePayload accepts base64 forms Go's StdEncoding accepts", () => {
  assert.equal(
    new TextDecoder().decode(decodeEnvelopePayload(JSON.stringify({ payload: "e3\n0=" }))),
    "{}",
  );
  assert.deepEqual(Buffer.from(decodeEnvelopePayload(JSON.stringify({ payload: "AB==" }))), Buffer.from([0]));
});

test("decodeEnvelopePayload rejects base64 forms Go's StdEncoding rejects", () => {
  for (const payload of ["q83v_-7dqw", "e30", "e30=@@@"]) {
    assert.throws(() => decodeEnvelopePayload(JSON.stringify({ payload })), EnvelopeFormatError);
  }
});

test("verify rejects a structurally malformed envelope with InvalidSignatureError, not a raw TypeError", () => {
  const ks = testKeyset();
  const keysets = [publicKeyset(ks)];

  const badPayloads = [
    { signature: { keysetId: "ks_test", signatures: ["AA=="] } }, // payload absent
    { payload: null, signature: { keysetId: "ks_test", signatures: ["AA=="] } },
    { payload: 1, signature: { keysetId: "ks_test", signatures: ["AA=="] } },
    { payload: {}, signature: { keysetId: "ks_test", signatures: ["AA=="] } },
    { payload: "AA==", signature: { keysetId: "ks_test", signatures: [null, "AA=="] } },
    { payload: "AA==", signature: { keysetId: "ks_test", signatures: [1, "AA=="] } },
    { payload: "AA==" }, // signature absent
    { payload: "AA==", signature: null },
    { payload: "AA==", signature: { keysetId: "ks_test", signatures: ["AA==", "AA=="] } },
  ];

  for (const bad of badPayloads) {
    assert.throws(
      () => verifyEnvelope(JSON.stringify(bad), keysets),
      InvalidSignatureError,
    );
  }

  // A resolvable keyset with keys: null must also reject cleanly.
  const nullKeysKeysets = [{ keysetId: "ks_test", keys: null }] as unknown as Parameters<
    typeof verifyEnvelope
  >[1];
  assert.throws(
    () =>
      verifyEnvelope(
        JSON.stringify({ payload: "AA==", signature: { keysetId: "ks_test", signatures: [] } }),
        nullKeysKeysets,
      ),
    InvalidSignatureError,
  );
});

test("decodeEnvelopePayload throws EnvelopeFormatError on malformed input, never InvalidSignatureError", () => {
  for (const bad of ['{', '{"payload":null}', '{"signature":{}}', "null"]) {
    assert.throws(() => decodeEnvelopePayload(bad), EnvelopeFormatError);
  }
});

// alg "none" (or any unrecognized algorithm) must fail closed.
test("verifyComponent rejects an unsupported algorithm", () => {
  assert.equal(
    verifyComponent("none", new Uint8Array(32), new Uint8Array([1]), new Uint8Array([2])),
    false,
  );
});

test("envelope pins the base64 alphabet and the JSON key names", () => {
  const ks = generateKeyset("ks_wire");
  // Encodes to "/++++Q==" in the standard alphabet and "_----Q==" in
  // base64url, so the two are distinguishable here where short ASCII
  // payloads make them identical.
  const payload = new Uint8Array([0xff, 0xef, 0xbe, 0xfb, 0xff, 0x41]);
  const want = Buffer.from(payload).toString("base64");

  const env = JSON.parse(signEnvelope(ks, payload));
  assert.deepEqual(Object.keys(env).sort(), ["payload", "signature"]);
  assert.deepEqual(Object.keys(env.signature).sort(), ["keysetId", "signatures"]);
  assert.equal(env.payload, want);
  assert.equal(env.signature.keysetId, "ks_wire");
  assert.equal(env.signature.signatures.length, KEYSET_ALGS.length);
});

test("ML-DSA signing is deterministic", () => {
  const ks = testKeyset();
  const mldsaKey = ks.keys.find((k) => k.alg === ALG_ML_DSA_65)!;
  const msg = new TextEncoder().encode("hello");

  const sig1 = signComponent(mldsaKey, msg);
  const sig2 = signComponent(mldsaKey, msg);

  assert.deepEqual(Buffer.from(sig1), Buffer.from(sig2));
});
